File: analytics/src/main/java/io/appmetrica/analytics/impl/AppStatusMonitor.java
package io.appmetrica.analytics.impl;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor;
import java.util.HashSet;
import java.util.Set;

public class AppStatusMonitor {

    public interface Observer {

        void onResume();

        void onPause();

    }

    @NonNull
    private final ICommonExecutor mExecutor;

    private final long mDefaultSessionTimeout;

    private final Set<ObserverWrapper> mObservers = new HashSet<ObserverWrapper>();

    private boolean mPaused = true;

    public AppStatusMonitor(long defaultSessionTimeout) {
        this(
                defaultSessionTimeout,
                ClientServiceLocator.getInstance().getClientExecutorProvider().getDefaultExecutor()
        );
    }

    AppStatusMonitor(long defaultSessionTimeout, @NonNull ICommonExecutor executor) {
        mExecutor = executor;
        mDefaultSessionTimeout = defaultSessionTimeout;
    }

    public synchronized void resume() {
        mPaused = false;
        for (ObserverWrapper observer : mObservers) {
            observer.notifyOnResume();
        }
    }

    public synchronized void pause() {
        mPaused = true;
        for (ObserverWrapper observer : mObservers) {
            observer.notifyOnPause();
        }
    }

    public synchronized void registerObserver(@NonNull Observer observer) {
        registerObserver(observer, false);
    }

    public synchronized void registerObserver(@NonNull Observer observer, boolean sticky) {
        registerObserver(observer, mDefaultSessionTimeout, sticky);
    }

    public synchronized void registerObserver(@NonNull Observer observer, long timeout) {
        registerObserver(observer, timeout, false);
    }

    public synchronized void registerObserver(@NonNull Observer observer, long timeout, boolean sticky) {
        ObserverWrapper observerWrapper = new ObserverWrapper(observer, mExecutor, timeout);
        mObservers.add(observerWrapper);
        // ObserverWrapper is paused by default and do not need to call notifyOnPause
        if (sticky && mPaused == false) {
            observerWrapper.notifyOnResume();
        }
    }

    public synchronized void unregisterObserver(@NonNull Observer observer) {
        ObserverWrapper observerToRemove = null;
        for (ObserverWrapper observerWrapper : mObservers) {
            if (observerWrapper.observer.equals(observer)) {
                observerToRemove = observerWrapper;
                break;
            }
        }
        mObservers.remove(observerToRemove);
    }

    private class ObserverWrapper {
        @NonNull
        final ICommonExecutor mExecutor;
        @NonNull
        final Observer observer;
        private final long mTimeout;

        private boolean mPaused = true;

        private final Runnable mPauseRunnable = new Runnable() {
            @Override
            public void run() {
                observer.onPause();
            }
        };

        ObserverWrapper(@NonNull Observer observer, @NonNull ICommonExecutor executor, long timeout) {
            this.observer = observer;
            mExecutor = executor;
            mTimeout = timeout;
        }

        void notifyOnResume() {
            if (mPaused) {
                mPaused = false;
                mExecutor.remove(mPauseRunnable);
                observer.onResume();
            }
        }

        void notifyOnPause() {
            if (mPaused == false) {
                mPaused = true;
                mExecutor.executeDelayed(mPauseRunnable, mTimeout);
            }
        }
    }

    @NonNull
    @VisibleForTesting
    ICommonExecutor getExecutor() {
        return mExecutor;
    }

    @VisibleForTesting
    long getDefaultSessionTimeout() {
        return mDefaultSessionTimeout;
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/AppStatusMonitor.kt
package io.appmetrica.analytics.impl

import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger

internal class AppStatusMonitor {

    internal interface Observer {
        fun onResume()
        fun onPause()
    }

    private val tag = "[AppStatusMonitor]"

    private val observers = mutableSetOf<ObserverWrapper>()
    private var paused = true

    @Synchronized
    fun resume() {
        DebugLogger.info(tag, "resume")
        paused = false
        for (observer in observers) {
            observer.notifyOnResume()
        }
    }

    @Synchronized
    fun pause() {
        DebugLogger.info(tag, "pause")
        paused = true
        for (observer in observers) {
            observer.notifyOnPause()
        }
    }

    @Synchronized
    fun registerObserver(observer: Observer, timeout: Long) {
        registerObserver(observer, timeout, false)
    }

    @Synchronized
    fun registerObserver(observer: Observer, timeout: Long, sticky: Boolean) {
        val observerWrapper = ObserverWrapper(
            observer,
            ClientServiceLocator.getInstance().clientExecutorProvider.defaultExecutor,
            timeout
        )
        observers.add(observerWrapper)
        // ObserverWrapper is paused by default and do not need to call notifyOnPause
        if (sticky && !paused) {
            observerWrapper.notifyOnResume()
        }
    }

    @Synchronized
    fun unregisterObserver(observer: Observer) {
        observers.removeAll { it.observer == observer }
    }

    private class ObserverWrapper(val observer: Observer, val executor: ICommonExecutor, val timeout: Long) {
        private val tag = "[AppStatusMonitor.ObserverWrapper]"

        private var paused = true

        private val pauseRunnable = Runnable {
            DebugLogger.info(tag, "Notify on pause")
            observer.onPause()
        }

        fun notifyOnResume() {
            if (paused) {
                paused = false
                executor.remove(pauseRunnable)
                DebugLogger.info(tag, "Notify on resume")
                observer.onResume()
            }
        }

        fun notifyOnPause() {
            if (!paused) {
                paused = true
                DebugLogger.info(tag, "Schedule pause runnable with %s seconds delay", timeout)
                executor.executeDelayed(pauseRunnable, timeout)
            }
        }
    }
}
